src/replication.c

```diff
@@ -2308,6 +2474,441 @@ char *sendCommandArgv(connection *conn, int argc, char **argv, size_t *argv_lens
+    serverLog(LL_WARNING, "Primary did not respond to command during SYNC handshake");
+    /* Fall through to regular error handling */
+
+error:
+    connClose(conn);
+    server.repl_transfer_s = NULL;
+    if (server.repl_rdb_transfer_s) {
+        connClose(server.repl_rdb_transfer_s);
+        server.repl_rdb_transfer_s = NULL;
+    }
+    if (server.repl_transfer_fd != -1) close(server.repl_transfer_fd);
+    server.repl_transfer_fd = -1;
+    server.repl_state = REPL_STATE_CONNECT;
+    abortRdbConnectionSync();
+    return;
+
+write_error: /* Handle sendCommand() errors. */
+    serverLog(LL_WARNING, "Sending command to primary in dual connection replication handshake: %s", err);
+    sdsfree(err);
+    goto error;
+}
+
+/* Replication: Replica side.
+ * Initialize server.pending_repl_data infrastructure, we will allocate the buffer
+ * itself once we need it */
+void replDataBufInit(void) {
+    serverAssert(server.pending_repl_data.blocks == NULL);
+    server.pending_repl_data.len = 0;
+    server.pending_repl_data.peak = 0;
+    server.pending_repl_data.blocks = listCreate();
+    server.pending_repl_data.blocks->free = zfree;
+}
+
+/* Replication: Replica side.
+ * Track the local repl-data buffer streaming progress and serve clients from time to time */
+void replStreamProgressCallback(size_t offset, int readlen, time_t *last_progress_callback) {
+    time_t now = mstime();
+    if (server.loading_process_events_interval_bytes &&
+        (offset + readlen) / server.loading_process_events_interval_bytes >
+            offset / server.loading_process_events_interval_bytes &&
+        now - *last_progress_callback > server.loading_process_events_interval_milliseconds) {
+        replicationSendNewlineToPrimary();
+        processEventsWhileBlocked();
+        *last_progress_callback = now;
+    }
+}
+
+/* Replication: Replica side.
+ * Reads replication data from primary into specified repl buffer block */
+int readIntoReplDataBlock(connection *conn, replDataBufBlock *o, size_t read) {
+    int nread = connRead(conn, o->buf + o->used, read);
+    if (nread == -1) {
+        if (connGetState(conn) != CONN_STATE_CONNECTED) {
+            serverLog(LL_VERBOSE, "Error reading from primary: %s", connGetLastError(conn));
+            cancelReplicationHandshake(1);
+        }
+        return C_ERR;
+    }
+    if (nread == 0) {
+        if (server.verbosity <= LL_VERBOSE) {
+            serverLog(LL_VERBOSE, "Provisional primary closed connection");
+        }
+        cancelReplicationHandshake(1);
+        return C_ERR;
+    }
+    o->used += nread;
+    server.stat_total_reads_processed += nread;
+    return read - nread;
+}
+
+/* Replication: Replica side.
+ * Returns true in case the replica's local repl-baffer used all of its space */
+int isReplicaBufferLimitReached(void) {
+    return server.pending_repl_data.len > server.client_obuf_limits[CLIENT_REPLICA].hard_limit_bytes;
```

Comment:
Yeah agreed that we do need to regulate the "producer" here. My comment though was on the policy we pick. We are relying on the replica's output buffer limit to regulate the primary's COB. If we were to use "maxmemory" in your case, we would arrive at 1GB as well. On the other hand though, I guess `client_obuf_limits[CLIENT_REPLICA]` works too if we think of it as a cluster wide config, since no one will config nodes in the same cluster with a different `client_obuf_limits[CLIENT_REPLICA]` value. I think I can live with this policy too for now in this sense.

Reply:
> If we were to use "maxmemory" in your case, we would arrive at 1GB as well. 

Not necessarily. It really depends on the RDB load speed. Consider a replica that loads the RDB slowly. For instance, at time t_0, the replica has loaded 0 bytes of the RDB but has accumulated 2GB of replication data. At this point, we can be certain that the sync will eventually fail because the replica only has 8GB of free space left to store an RDB that is 9GB in size.

Reply:
Good point!